lib/vscode/src/vs/server/net/webRequestHandler.ts

```diff
@@ -0,0 +1,329 @@
+
+		if (knownCallback) {
+			this.callbackUriToRequestId.delete(requestId);
+			clearTimeout(knownCallback.timeout);
+		}
+
+		res.writeHead(200, { 'Content-Type': 'text/json' });
+		return res.end(JSON.stringify(knownCallback?.uri));
+	};
+
+	/**
+	 * Remote resource endpoint.
+	 * @remark Used to load resources on the client-side.
+	 */
+	private $remoteResource: WebRequestListener = async (req, res) => {
+		const path = req.parsedUrl.searchParams.get('path');
+
+		if (path) {
+			res.setHeader('Content-Type', getMediaMime(path));
+			res.end(await fs.readFile(path));
+		}
+	};
+
+	// /**
+	//  * Webview endpoint
+	//  */
+	// private $webview: WebRequestListener = async (req, res) => {
+	// 	const foo = req.parsedUrl.pathname.foo;
+
+	// 	if (/^vscode-resource/.test(foo)) {
+	// 		return serveFile(req, res, foo.replace(/^vscode-resource(\/file)?/, ''));
+	// 	}
+
+	// 	return serveFile(req, res, join(paths.WEBVIEW, foo));
+	// };
+
+	public override dispose() {
+		super.dispose();
+		this.callbackUriToRequestId.clear();
+	}
+}
+
+async function serveError(res: ServerResponse, errorCode: number, errorMessage: string, responseHeaders = Object.create(null)) {
+	responseHeaders['Content-Type'] = 'text/plain';
+	res.writeHead(errorCode, responseHeaders);
+	res.end(errorMessage);
+}
+
+async function serveFile(req: IncomingMessage, res: ServerResponse, filePath: string, responseHeaders = Object.create(null)) {
+	try {
+		// Sanity checks
+		filePath = normalize(filePath); // ensure no "." and ".."
+
+		const stat = await fs.stat(filePath);
+
+		// Check if file modified since
+		// Weak validator (https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag)
+		const etag = `W/"${[stat.ino, stat.size, stat.mtime.getTime()].join('-')}"`;
+		if (req.headers['if-none-match'] === etag) {
+			res.writeHead(304);
+			return res.end();
+		}
+
+		// Headers
+		responseHeaders['Content-Type'] = textMimeType[extname(filePath)] || getMediaMime(filePath);
+		responseHeaders['Etag'] = etag;
+
+		res.writeHead(200, responseHeaders);
+
+		// Data
+		createReadStream(filePath).pipe(res);
+	} catch (error: any) {
+		console.error(error.toString());
+		responseHeaders['Content-Type'] = 'text/plain';
+		res.writeHead(404, responseHeaders);
+		return res.end('Not found');
+	}
+}
+
+const textMimeType: { [fileExt: string]: string | undefined } = {
+	'.html': 'text/html',
+	'.js': 'text/javascript',
+	'.json': 'application/json',
+	'.css': 'text/css',
+	'.svg': 'image/svg+xml',
+};
+
+const mapExtToMediaMimes: { [fileExt: string]: string | undefined } = {
+	'.bmp': 'image/bmp',
+	'.gif': 'image/gif',
+	'.ico': 'image/x-icon',
+	'.jpe': 'image/jpg',
+	'.jpeg': 'image/jpg',
+	'.jpg': 'image/jpg',
+	'.png': 'image/png',
+	'.tga': 'image/x-tga',
+	'.tif': 'image/tiff',
+	'.tiff': 'image/tiff',
+	'.woff': 'application/font-woff',
+};
+
+function getMediaMime(forPath: string) {
+	const ext = extname(forPath);
+
+	return mapExtToMediaMimes[ext.toLowerCase()] || 'text/plain';
```

Comment:
This should check for `textMimeType` too... otherwise `.svg` won't be properly rendered in the browser.